zenith_utils/src/pq_proto.rs

```diff
@@ -169,44 +169,58 @@ impl FeStartupMessage {
             bail!("invalid message length");
         }
 
-        let version = stream.read_u32::<BE>()?;
-        let kind = match version {
-            CANCEL_REQUEST_CODE => StartupRequestCode::Cancel,
-            NEGOTIATE_SSL_CODE => StartupRequestCode::NegotiateSsl,
-            NEGOTIATE_GSS_CODE => StartupRequestCode::NegotiateGss,
-            _ => StartupRequestCode::Normal,
-        };
+        let request_code = stream.read_u32::<BE>()?;
 
         // the rest of startup packet are params
         let params_len = len - 8;
         let mut params_bytes = vec![0u8; params_len];
         stream.read_exact(params_bytes.as_mut())?;
 
-        // Then null-terminated (String) pairs of param name / param value go.
-        let params_str = str::from_utf8(&params_bytes).unwrap();
-        let params = params_str.split('\0');
-        let mut params_hash: HashMap<String, String> = HashMap::new();
-        for pair in params.collect::<Vec<_>>().chunks_exact(2) {
-            let name = pair[0];
-            let value = pair[1];
-            if name == "options" {
-                // deprecated way of passing params as cmd line args
-                for cmdopt in value.split(' ') {
-                    let nameval: Vec<&str> = cmdopt.split('=').collect();
-                    if nameval.len() == 2 {
-                        params_hash.insert(nameval[0].to_string(), nameval[1].to_string());
+        // Parse params depending on request code
+        let most_sig_16_bits = request_code >> 16;
+        let least_sig_16_bits = request_code & ((1 << 16) - 1);
+        let message = match (most_sig_16_bits, least_sig_16_bits) {
+            (RESERVED_INVALID_MAJOR_VERSION, CANCEL_REQUEST_CODE) => {
+                ensure!(params_len == 8, "expected 8 bytes for CancelRequest params");
+                FeInitialMessage::CancelRequest {
+                    backend_pid: byteorder::BigEndian::read_i32(&params_bytes[0..4]),
+                    cancel_key: byteorder::BigEndian::read_i32(&params_bytes[4..8]),
+                }
```

Comment:
I think you might be able to make this cleaner with something like

```
let cursor = Cursor::new(param_bytes);
cursor.read_i32::<BE>()
```

I'm not 100% sure on that exact syntax though